Add Topup BiasFB

fernandohds564 opened the pull request:
This PR main feature is the addition of a Bias Feedback to the InjCtrl IOC for topup operation.
The logic of the feedback was inspired in the class added in lnls-fac/apsuite#201. 

Besides this addition, a few other classes were changed:
 - Callback: I added a property called `has_callbacks`.
 - CurrInfoSI: the InjCurr PV will always be updated when Current3GeV from the booster is updated.
 - CurrInfoSI, Lifetime and InjCtrl: the increment of the `-Cmd` was delegated to the Driver class of the IOC. See lnls-sirius/machine-applications#266

xresende commented on siriuspy/siriuspy/injctrl/csdev.py:
```diff
@@ -289,3 +297,192 @@ def get_injctrl_propty_database():
+            'prec': 1, 'lolim': -120, 'low': -120, 'lolo': -120,
+            'hilim': -30.0, 'high': -30.0, 'hihi': -30.0},
+
+        'ModelType-Sel': {
```
we should add units for similar PVs (`ModelMaxNumPts-SP/RB`, for example) in order for archiver viewer to be used more easily.

fernandohds564 replied:
I don't understand @xresende. I have added units for this PV you commented. I just didn't add for `-Sel/-Sts` PVs. Do you think I should add for them too?

xresende replied:
I dont understand either. I meant `ModelMaxNumPts-SP/RB`, as I wrote. But right now I can see on this github.com PR interface that the units are there! I remember looking at the pair `ModelMaxNumPts-SP/RB` and not finding unit in neither of them. And you did not commit any changes... Well, most likely I overlooked something... sorry.

As to `Sel/Sts` PVs, I think we should also add units. Specially if the property is such that other devices might have identical or similar properties. Besides, adding units  you avoid archiver viewer having one axis for `Sel`and another one for `Sts`. I would suggest we copy the name of the property (`ModelType`, for example) to the PV unit field.

fernandohds564 replied:
ok! I can do that!

fernandohds564 replied:
@xresende, I've done that. I'll explain how I did it in the main thread so it is more visible to others.

murilobalves commented on siriuspy/siriuspy/currinfo/main.py:
```diff
@@ -611,17 +611,12 @@ def _update_injeff(self):
+        self._injeff = 0.0
+        # check if buffer not empty and if there is valid current in Booster
+        if timestamp_dq.size and bo_curr >= self.CURR_THRESHOLD:
+            self._injcurr = value_dq.max() - value_dq.min()  # mA
```
you can do the same with `value_dq.ptp()` to get peak-to-peak value, if you want

fernandohds564 replied:
Nice! Done, thanks!

fernandohds564 commented:
@xresende, as you requested, I added units to the `-Sel/Sts` and also to the `-Mon` PVs which are of type `enum`. I also added units to `int` PVs which are arrays of booleans to indicate interlock or ok/not ok states.

However I didn't follow your suggestion of copying the name of the PV to the unit. I thought it would be more descriptive to summarize all the possible states meaning in order, so the user know how to translate the numbers. For example, the `Off/On` PVs have unit `Off_On`.

The `int` PVs which are arrays of booleans follow the same idea, but the unit has a `2^_` in front of them to indicate that the value of the PV is a sum of powers of two.

@anacso17 , I tried to propagate this idea to the other PVs of the InjCtrl IOC. Please let me know what is your opinion about it.
